lib/Dialect/FIRRTL/Transforms/LowerBundleVectorTypes.cpp

```diff
@@ -0,0 +1,958 @@
+//===- LowerBundleVectorTypes.cpp - Lower Aggregate Types -------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the LowerBundleVectorTypes pass.  This pass replaces
+// aggregate types with expanded values.
+//
+// This pass walks the operations in reverse order. This lets it visit users
+// before defs. Users can usually be expanded out to multiple operations (think
+// mux of a bundle to muxes of each field) with a temporary subWhatever op
+// inserted. When processing an aggregate producer, we blow out the op as
+// appropriate, then walk the users, often those are subWhatever ops which can
+// be bypassed and deleted. Function arguments are logically last on the
+// operation visit order and walked left to right, being peeled one layer at a
+// time with replacements inserted to the right of the original argument.
+//
+// Each processing of an op peels one layer of aggregate type off.  Because new
+// ops are inserted immediately above the current up, the walk will visit them
+// next, effectively recusing on the aggregate types, without recusing.  These
+// potentially temporary ops(if the aggregate is complex) effectively serve as
+// the worklist.  Often aggregates are shallow, so the new ops are the final
+// ones.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLAttributes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/Support/Parallel.h"
+#include <deque>
+
+using namespace circt;
+using namespace firrtl;
+
+// TODO: check all argument types
+namespace {
+/// This represents a flattened bundle field element.
+struct FlatBundleFieldEntry {
+  /// This is the underlying ground type of the field.
+  FIRRTLType type;
+  /// The index in the parent type
+  size_t index;
+  /// The FieldID range of self and children in the parent type
+  size_t fieldID, fieldIDWidth;
```

Comment:
Would it make sense to align with `subAnnotationAttr` here?
```suggestion
  size_t minFieldID, maxFieldID;
```

Reply:
I went back and forth, but went with your suggestion.